Test regressions support

ana opened the pull request:
The following changes store the test regressions in the backend and update the test report mail. The implementation is using the same strategy than the boot regressions and some functions should be moved to a common library.
Some notes: 
* It doesn't work with `/test`,  only when adding test via the LAVA callback
* This implementation uses the test case name to report about the regressions, if a test group have two test cases with the same name, it'll lead to false positives. (Two test cases with same name in different test groups is OK)
* Having sub-groups with the same name from different groups or a subgroup with the name of a top group can a also lead to bad results.

A mail with the regressions will look like this:

```
Subject: Test results (simple) for local/master - v4.20-rc4-301-g13294fbcd4c1

Test results for:
  Tree:    local
  Branch:  master
  Kernel:  v4.20-rc4-301-g13294fbcd4c1
  URL:     /home/ana/KernelCI/Git/linux
  Commit:  13294fbcd4c195382a0c708c424cd36e49ff686f
  Test plans: simple

Summary
-------
1 test groups results

1  | simple     | qemu                   | arm64 |   8 total:   3 PASS   5 FAIL   0 SKIP

---
1  | simple     | qemu                   | arm64 |   8 total:   3 PASS   5 FAIL   0 SKIP

  Config:      defconfig
  Lab Name:    lab-ana
  Date:        2018-12-05 13:59:48.909000
  TXT log:     http://storage//local/master/v4.20-rc4-301-g13294fbcd4c1/arm64/defconfig/lab-ana/simple-qemu.txt
  HTML log:    http://storage//local/master/v4.20-rc4-301-g13294fbcd4c1/arm64/defconfig/lab-ana/simple-qemu.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch/20181122.0/arm64/rootfs.cpio.gz

    * ls: last worked in v4.20-rc4-299-gb6839ef26e54

    networking - 3 tests: 2  PASS, 1 FAIL, 0 SKIP

      * ls: last worked in v4.20-rc4-299-gb6839ef26e54

    filesystem - 3 tests: 0  PASS, 3 FAIL, 0 SKIP

      * df: last worked in v4.20-rc4-299-gb6839ef26e54
      * ls: last worked in v4.20-rc4-300-g34caf313b358
      * mount: last worked in v4.20-rc4-300-g34caf313b358

```


Some things to do before merging this:
* Add indexes for the two new collections
* Add unit tests for `/app/utils/kci_test/regressions.py`
* Rebase in the top of other changes (most notably PR #96)

gctucker commented:
Looks like good progress :)

Regarding things that have always failed, I was thinking we may need to have a way to set some initial condition to start regression tracking only for tests that are meant to pass.  Otherwise, tests that keep failing will never appear in the summary if we're strictly only showing regressions.

Or, based on the example above where we're showing all failures even if they're not regressions, we could have something like this:
```
* whoami: never passed
```
One issue with this approach is with very large test suites with tests that will never be fixed, causing a lot of noise.  We don't have this problem yet with our current test suites but it might become a real issue.  Maybe we could add a blacklist mechanism later on to not show some failures in the report, I think that could also be defined in `test-configs.yaml`.

ana commented:
I just updated to the v3, this needs a new run of tests in staging first, but I think it's ready to merge.

Changes:
* Updated message strings in the report
* Handle some cases that appear during the previous round of testing in staging (like handle when a test case goes from SKIP to FAIL)
* Add unit tests for the `/app/utils/kci_test/regressions.py`

Example mail, with two regressions and a test that when from SKIP to FAIL. It's not included in this example when a test always failed since it's been tracked, they will show as "never passed".
```
Subject | Test results for gtucker/media-ana - gtucker-kernelci-001-30753-g8999e22a8cdd

Test results for:
  Tree:    gtucker
  Branch:  media-ana
  Kernel:  gtucker-kernelci-001-30753-g8999e22a8cdd
  URL:     /home/collabora/KernelCI/Git/media-gtucker
  Commit:  8999e22a8cdd10c7f3b696e99f02d48a14d0aa16
  Test plans: All the results are included

Summary
-------
1 test groups results

1  | v4l2       | qemu                   | arm64 |  48 total:  26 PASS   3 FAIL  19 SKIP

---
1  | v4l2       | qemu                   | arm64 |  48 total:  26 PASS   3 FAIL  19 SKIP

  Config:      defconfig+virtualvideo
  Lab Name:    lab-ana
  Date:        2018-12-17 06:54:32.982000
  TXT log:     http://storage//gtucker/media-ana/gtucker-kernelci-001-30753-g8999e22a8cdd/arm64/defconfig+virtualvideo/lab-ana/v4l2-qemu.txt
  HTML log:    http://storage//gtucker/media-ana/gtucker-kernelci-001-30753-g8999e22a8cdd/arm64/defconfig+virtualvideo/lab-ana/v4l2-qemu.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch-v4l2/20181207.0/arm64/rootfs.cpio.gz
  Test Git:    git://linuxtv.org/v4l-utils.git
  Test Commit: 7a118f166609c0d05c6447cc79484af37875d6fc



    Input/Output-configuration-ioctls - 4 tests: 0  PASS, 0 FAIL, 4 SKIP
    
  
    Debug-ioctls - 2 tests: 1  PASS, 0 FAIL, 1 SKIP
    
  
    Codec-ioctls-Input-0 - 3 tests: 0  PASS, 0 FAIL, 3 SKIP
    
  
    Input-ioctls - 6 tests: 2  PASS, 2 FAIL, 2 SKIP
    
      * VIDIOC_ENUMAUDIO: last worked in gtucker-kernelci-001-30752-g9f53f1678fad
      * VIDIOC_G/S/ENUMINPUT: last worked in gtucker-kernelci-001-30752-g9f53f1678fad
  
    Output-ioctls - 5 tests: 1  PASS, 1 FAIL, 3 SKIP
    
      * VIDIOC_G/S_AUDOUT: new FAIL, skipped in gtucker-kernelci-001-30752-g9f53f1678fad
  
    Format-ioctls-Input-0 - 10 tests: 6  PASS, 0 FAIL, 4 SKIP
    
  
    Buffer-ioctls-Input-0 - 3 tests: 2  PASS, 0 FAIL, 1 SKIP
    
  
    Streaming-ioctls_Test-input-0 - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Allow-for-multiple-opens - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Required-ioctls - 1 tests: 1  PASS, 0 FAIL, 0 SKIP
    
  
    Control-ioctls-Input-0 - 6 tests: 5  PASS, 0 FAIL, 1 SKIP
```

gctucker commented:
Great, thanks.  I'll spin a new round with this on staging.

gctucker commented:
Got a report showing some regressions caused on purpose:
```
    Input-ioctls - 6 tests: 2  PASS, 2 FAIL, 2 SKIP
    
      * VIDIOC_ENUMAUDIO: last worked in v4.20-rc7-268-ga87d2f0cba3d
      * VIDIOC_G/S/ENUMINPUT: last worked in v4.20-rc7-268-ga87d2f0cba3d
```
Full report, showing half-regressions for things that never passed or were skipped before:
```
Test results for:
  Tree:    gtucker
  Branch:  kernelci-media
  Kernel:  v4.20-rc7-277-g87660b415e49
  URL:     https://gitlab.collabora.com/gtucker/linux.git
  Commit:  87660b415e4916f0fee4130a502a3074bbd9bf94
  Test plans: v4l2

Summary
-------
4 test groups results

1  | v4l2       | rk3399-gru-kevin       | arm64 |  49 total:  17 PASS   6 FAIL  26 SKIP
2  | v4l2       | rk3288-veyron-jaq      | arm   |  49 total:  17 PASS   6 FAIL  26 SKIP
3  | v4l2       | qemu                   | arm   |  49 total:  27 PASS   4 FAIL  18 SKIP
4  | v4l2       | qemu                   | arm64 |  49 total:  27 PASS   4 FAIL  18 SKIP

---
1  | v4l2       | rk3399-gru-kevin       | arm64 |  49 total:  17 PASS   6 FAIL  26 SKIP

  Config:      defconfig
  Lab Name:    lab-collabora-dev
  Date:        2018-12-18 18:41:56.709000
  TXT log:     http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm64/defconfig/lab-collabora-dev/v4l2-rk3399-gru-kevin.txt
  HTML log:    http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm64/defconfig/lab-collabora-dev/v4l2-rk3399-gru-kevin.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch-v4l2/20181207.0/arm64/rootfs.cpio.gz
  Test Git:    git://linuxtv.org/v4l-utils.git
  Test Commit: 7a118f166609c0d05c6447cc79484af37875d6fc

  

    Codec-ioctls-Input-0 - 3 tests: 0  PASS, 0 FAIL, 3 SKIP
    
  
    Input/Output-configuration-ioctls - 4 tests: 0  PASS, 0 FAIL, 4 SKIP
    
  
    Output-ioctls - 5 tests: 0  PASS, 0 FAIL, 5 SKIP
    
  
    Required-ioctls - 2 tests: 2  PASS, 0 FAIL, 0 SKIP
    
  
    Allow-for-multiple-opens - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Control-ioctls-Input-0 - 6 tests: 3  PASS, 2 FAIL, 1 SKIP
    
      * VIDIOC_G/S/TRY_EXT_CTRLS: never passed
      * VIDIOC_QUERY_EXT_CTRL/QUERYMENU: never passed
  
    Streaming-ioctls_Test-input-0 - 4 tests: 1  PASS, 2 FAIL, 1 SKIP
    
      * USERPTR: never passed
      * MMAP: never passed
  
    Debug-ioctls - 2 tests: 0  PASS, 0 FAIL, 2 SKIP
    
  
    Format-ioctls-Input-0 - 10 tests: 4  PASS, 1 FAIL, 5 SKIP
    
      * VIDIOC_G/S_PARM: never passed
  
    Buffer-ioctls-Input-0 - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
    
  
    Input-ioctls - 6 tests: 1  PASS, 0 FAIL, 5 SKIP
    
  
2  | v4l2       | rk3288-veyron-jaq      | arm   |  49 total:  17 PASS   6 FAIL  26 SKIP

  Config:      multi_v7_defconfig
  Lab Name:    lab-collabora-dev
  Date:        2018-12-18 18:43:32.950000
  TXT log:     http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm/multi_v7_defconfig/lab-collabora-dev/v4l2-rk3288-veyron-jaq.txt
  HTML log:    http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm/multi_v7_defconfig/lab-collabora-dev/v4l2-rk3288-veyron-jaq.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch-v4l2/20181207.0/armhf/rootfs.cpio.gz
  Test Git:    git://linuxtv.org/v4l-utils.git
  Test Commit: 7a118f166609c0d05c6447cc79484af37875d6fc

  

    Codec-ioctls-Input-0 - 3 tests: 0  PASS, 0 FAIL, 3 SKIP
    
  
    Input/Output-configuration-ioctls - 4 tests: 0  PASS, 0 FAIL, 4 SKIP
    
  
    Output-ioctls - 5 tests: 0  PASS, 0 FAIL, 5 SKIP
    
  
    Required-ioctls - 2 tests: 2  PASS, 0 FAIL, 0 SKIP
    
  
    Allow-for-multiple-opens - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Control-ioctls-Input-0 - 6 tests: 3  PASS, 2 FAIL, 1 SKIP
    
      * VIDIOC_G/S/TRY_EXT_CTRLS: never passed
      * VIDIOC_QUERY_EXT_CTRL/QUERYMENU: never passed
  
    Streaming-ioctls_Test-input-0 - 4 tests: 1  PASS, 2 FAIL, 1 SKIP
    
      * USERPTR: never passed
      * MMAP: never passed
  
    Debug-ioctls - 2 tests: 0  PASS, 0 FAIL, 2 SKIP
    
  
    Format-ioctls-Input-0 - 10 tests: 4  PASS, 1 FAIL, 5 SKIP
    
      * VIDIOC_G/S_PARM: never passed
  
    Buffer-ioctls-Input-0 - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
    
  
    Input-ioctls - 6 tests: 1  PASS, 0 FAIL, 5 SKIP
    
  
3  | v4l2       | qemu                   | arm   |  49 total:  27 PASS   4 FAIL  18 SKIP

  Config:      multi_v7_defconfig+virtualvideo
  Lab Name:    lab-collabora-dev
  Date:        2018-12-18 19:04:38.544000
  TXT log:     http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm/multi_v7_defconfig+virtualvideo/lab-collabora-dev/v4l2-qemu.txt
  HTML log:    http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm/multi_v7_defconfig+virtualvideo/lab-collabora-dev/v4l2-qemu.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch-v4l2/20181207.0/armhf/rootfs.cpio.gz
  Test Git:    git://linuxtv.org/v4l-utils.git
  Test Commit: 7a118f166609c0d05c6447cc79484af37875d6fc



    Codec-ioctls-Input-0 - 3 tests: 0  PASS, 0 FAIL, 3 SKIP
    
  
    Input/Output-configuration-ioctls - 4 tests: 0  PASS, 0 FAIL, 4 SKIP
    
  
    Output-ioctls - 5 tests: 1  PASS, 1 FAIL, 3 SKIP
    
      * VIDIOC_G/S_AUDOUT: new FAIL, skipped in v4.20-rc7-268-ga87d2f0cba3d
  
    Required-ioctls - 2 tests: 2  PASS, 0 FAIL, 0 SKIP
    
  
    Allow-for-multiple-opens - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Control-ioctls-Input-0 - 6 tests: 5  PASS, 0 FAIL, 1 SKIP
    
  
    Streaming-ioctls_Test-input-0 - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Debug-ioctls - 2 tests: 1  PASS, 0 FAIL, 1 SKIP
    
  
    Format-ioctls-Input-0 - 10 tests: 6  PASS, 0 FAIL, 4 SKIP
    
  
    Buffer-ioctls-Input-0 - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
    
      * Requests: last worked in v4.20-rc5-279-ge159b6074c82
  
    Input-ioctls - 6 tests: 2  PASS, 2 FAIL, 2 SKIP
    
      * VIDIOC_ENUMAUDIO: last worked in v4.20-rc7-268-ga87d2f0cba3d
      * VIDIOC_G/S/ENUMINPUT: last worked in v4.20-rc7-268-ga87d2f0cba3d
  
4  | v4l2       | qemu                   | arm64 |  49 total:  27 PASS   4 FAIL  18 SKIP

  Config:      defconfig+virtualvideo
  Lab Name:    lab-collabora-dev
  Date:        2018-12-18 19:09:40.856000
  TXT log:     http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm64/defconfig+virtualvideo/lab-collabora-dev/v4l2-qemu.txt
  HTML log:    http://staging-storage.kernelci.org/gtucker/kernelci-media/v4.20-rc7-277-g87660b415e49/arm64/defconfig+virtualvideo/lab-collabora-dev/v4l2-qemu.html
  Rootfs:      http://storage.kernelci.org/images/rootfs/debian/stretch-v4l2/20181207.0/arm64/rootfs.cpio.gz
  Test Git:    git://linuxtv.org/v4l-utils.git
  Test Commit: 7a118f166609c0d05c6447cc79484af37875d6fc



    Codec-ioctls-Input-0 - 3 tests: 0  PASS, 0 FAIL, 3 SKIP
    
  
    Input/Output-configuration-ioctls - 4 tests: 0  PASS, 0 FAIL, 4 SKIP
    
  
    Output-ioctls - 5 tests: 1  PASS, 1 FAIL, 3 SKIP
    
      * VIDIOC_G/S_AUDOUT: new FAIL, skipped in v4.20-rc7-268-ga87d2f0cba3d
  
    Required-ioctls - 2 tests: 2  PASS, 0 FAIL, 0 SKIP
    
  
    Allow-for-multiple-opens - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Control-ioctls-Input-0 - 6 tests: 5  PASS, 0 FAIL, 1 SKIP
    
  
    Streaming-ioctls_Test-input-0 - 4 tests: 4  PASS, 0 FAIL, 0 SKIP
    
  
    Debug-ioctls - 2 tests: 1  PASS, 0 FAIL, 1 SKIP
    
  
    Format-ioctls-Input-0 - 10 tests: 6  PASS, 0 FAIL, 4 SKIP
    
  
    Buffer-ioctls-Input-0 - 3 tests: 2  PASS, 1 FAIL, 0 SKIP
    
      * Requests: last worked in v4.20-rc5-279-ge159b6074c82
  
    Input-ioctls - 6 tests: 2  PASS, 2 FAIL, 2 SKIP
    
      * VIDIOC_ENUMAUDIO: last worked in v4.20-rc7-268-ga87d2f0cba3d
      * VIDIOC_G/S/ENUMINPUT: last worked in v4.20-rc7-268-ga87d2f0cba3d
```

gctucker commented on app/handlers/dbindexes.py:
```diff
@@ -320,6 +321,33 @@ def _ensure_test_group_indexes(database):
     )
 
 
+def _ensure_test_group_regressions_indexes(database):
```
I thought this was supposed to be in #98? Can you rebase without these index patches, and make sure #98 is up-to-date so we can merge both PRs?

ana replied:
This two new collections are added in this PR (see the first commit), while #98 is for fixing the missing boot regressions indexes.

gctucker replied:
Ah yes sorry, my bad.  So I'll check #98 again and merge it if it looks ready as it's been tested already.

gctucker commented on app/utils/report/test.py:
```diff
@@ -56,7 +56,118 @@
+                post_case = _get_case(post["test_cases"], tc["name"])
+                prev_case = _get_case(prev["test_cases"], tc["name"])
+
+                if prev_case["status"] in ['PASS'] and post_case["status"] == 'FAIL':
```
How about `prev_case["status"] == 'PASS'`?

gctucker commented on app/utils/report/test.py:
```diff
@@ -56,7 +56,118 @@
+                    regr_dict.update(d)
+                    continue
+
+                if prev_case["status"] in ['SKIP'] and post_case["status"] == 'FAIL':
```
How about `prev_case["status"] == 'SKIP'`?

gctucker commented on app/utils/report/test.py:
```diff
@@ -56,7 +56,118 @@
+
+
 def _add_test_group_data(group, database):
+
```
Whitespace noise...

gctucker commented on app/utils/report/test.py:
```diff
@@ -81,11 +192,24 @@ def _add_test_group_data(group, database):
         "sub_groups": sub_groups,
         "total_tests": sum(total.values()),
         "total": total,
+        "regressions": regressions
```
Missing trailing comma.

gctucker commented on app/utils/report/test.py:
```diff
@@ -56,7 +56,118 @@
+                for instance_d in board_d.itervalues():
+                    for defconfig, defconfig_d in instance_d.iteritems():
+                        for compiler, tests in defconfig_d.iteritems():
+
```
Whitespace noise...

gctucker commented on app/utils/report/test.py:
```diff
@@ -56,7 +56,118 @@
+    if ndocs == 1:
+        # This should never happen
+        utils.LOG.warn("This message never should appear. There is \
+                        something wrong with the regression document!")
```
Might be worth throwing an exception here?

ana replied:
I don't know if I should leave these lines even. Every time the backend write a regression document, there will be at least two documents already.

gctucker commented on app/utils/kci_test/regressions.py:
```diff
@@ -0,0 +1,466 @@
+# How the key to access the regressions data structure is formatted.
+# Done in this way so that we can use mongodb dot notation to access embedded
+# documents.
+REGRESSION_FMT = "{:s}.{:s}.{:s}.{:s}.{:s}.{:s}"
```
Having named elements would be a bit more robust and self-documenting: `"{lab:s}.{arch:s}.`.

ana replied:
I haven't checked but  that'll be problematic if the different functions using REGRESSION_FMT use different parameters. This comes from `app/utils/boot/regressions.py` and eventually we would want to use the same format for both kind of regressions.

gctucker commented on app/utils/kci_test/regressions.py:
```diff
@@ -0,0 +1,466 @@
+    return ret_val, doc_id
+
+
+def _add_test_group_data(group, database):
```
Err, is this function not already in `app/utils/report/test.py`?  Also, why is it needed to update some test group data to while searching for regressions?

gctucker replied:
OK sounds like something to consolidate in a follow-up PR.

gctucker commented on app/utils/kci_test/regressions.py:
```diff
@@ -0,0 +1,466 @@
+
+
+def check_and_track(test_group_doc, db_options):
+    """Check previous test group report and start tracking regressions.
```
I think it's worth mentioning that this function doesn't recursively check for regressions in sub-groups, so the caller has to do the job.

gctucker replied:
It looks like this `check_and_track` function only tracks regressions for the test cases inside the given `test_group_doc`, but it doesn't look for sub-groups inside it.  So to track all regressions, this function must be called again for each sub-group and their own sub-groups recursively.  An alternative implementation would be to do the recursion in `check_and_track` so it would only need to be called once.

ana commented:
I pushed an updated addressing the comments above.

gctucker commented:
There are pep8 and unittest errors.  I'll test it on staging today anyway to check it's still working with the latest changes, but these errors will need to be fixed before we can merge the PR.

gctucker commented:
@ana Small problem, when I last ran this I got the regression as expected:
```
    Input-ioctls - 6 tests: 2  PASS, 2 FAIL, 2 SKIP
    
      * VIDIOC_ENUMAUDIO: last worked in v4.20-rc7-268-ga87d2f0cba3d
      * VIDIOC_G/S/ENUMINPUT: last worked in v4.20-rc7-268-ga87d2f0cba3d
```
However, this time with an extra patch on the same branch, the regression is lost:
```
    Input-ioctls - 6 tests: 2  PASS, 2 FAIL, 2 SKIP
    
      * VIDIOC_G/S/ENUMINPUT: never passed
      * VIDIOC_ENUMAUDIO: never passed
```
Do you think this can in any way be related to the build-environment changes in #86 ?  Maybe the regressions are broken because of the new build-environment meta-data, or should this have no effect at all?
cc @mattface ^

Meanwhile I'll try to see if boot regressions are still working.

ana commented:
I updated to fix the unittest problem (they actually caught a real issue!) .
There are some pep8 issues left from old changes, but I have also left some 3  ones of long lines because I don't want to fix that in a hurry without testing properly since they are the kind where you actually can break the code logic...

gctucker commented:
Replaced with #103.
